odoo/addons/base/models/ir_ui_menu.py

```diff
@@ -74,62 +74,69 @@ def _check_parent_id(self):
+            ['|', ('groups_id', '=', False), ('groups_id', 'in', tuple(group_ids))],
+            ['parent_id', 'action'], order='id',
+        ).sudo()
 
         # take apart menus that have an action
-        actions_by_model = defaultdict(set)
+        action_ids_by_model = defaultdict(list)
         for action in menus.mapped('action'):
             if action:
-                actions_by_model[action._name].add(action.id)
-        existing_actions = {
-            action
-            for model_name, action_ids in actions_by_model.items()
-            for action in self.env[model_name].browse(action_ids).exists()
-        }
-        action_menus = menus.filtered(lambda m: m.action and m.action in existing_actions)
-        folder_menus = menus - action_menus
-        visible = self.browse()
+                action_ids_by_model[action._name].append(action.id)
 
-        # process action menus, check whether their action is allowed
-        access = self.env['ir.model.access']
         MODEL_BY_TYPE = {
             'ir.actions.act_window': 'res_model',
             'ir.actions.report': 'model',
             'ir.actions.server': 'model_name',
         }
+        def exists_actions(model_name, action_ids):
+            """ Return existing actions and fetch model name field if exists"""
+            if model_name not in MODEL_BY_TYPE:
+                return self.env[model_name].browse(action_ids).exists()
+            records = self.env[model_name].sudo().with_context(active_test=False).search_fetch(
+                [('id', 'in', action_ids)], [MODEL_BY_TYPE[model_name]], order='id',
+            )
+            if model_name == 'ir.actions.server':
+                # Because it is computed, `search_fetch` doesn't fill the cache for it
+                records.mapped('model_name')
+            return records
 
-        # performance trick: determine the ids to prefetch by type
-        prefetch_ids = defaultdict(list)
-        for action in action_menus.mapped('action'):
-            prefetch_ids[action._name].append(action.id)
+        existing_actions = {
+            action
+            for model_name, action_ids in action_ids_by_model.items()
+            for action in exists_actions(model_name, action_ids)
+        }
+        action_menus = menus.filtered(lambda m: m.action in existing_actions)
+        folder_menu_ids = set((menus - action_menus)._ids)
 
+        # process action menus, check whether their action is allowed
+        access = self.env['ir.model.access']
+        visible_ids = set()
         for menu in action_menus:
             action = menu.action
-            action = action.with_prefetch(prefetch_ids[action._name])
+            # action[MODEL_BY_TYPE[action._name]] has been fetched in `exists_actions`
             model_name = action._name in MODEL_BY_TYPE and action[MODEL_BY_TYPE[action._name]]
             if not model_name or access.check(model_name, 'read', False):
```

Comment:
```suggestion
            fname = MODEL_BY_TYPE.get(action._name)
            if not fname:
                continue
            model_name = action[fname]
            if access.check(model_name, 'read', False):
```

Reply:
Mhhh it is not the same, If you don't have `fname`, you cannot check the access but you always add the menu to the visible one (see `not model_name or access.check(model_name, 'read', False)`). But it is a good idea to use `get` here (I will try both case to see which one is cleaner and faster).